### PR TITLE
MC/CUDA: fp16 reduce

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Added support for UCC_DT_FLOAT16 in cuda memory component

### CODE REVIEW

File: src/components/mc/cuda/kernel/mc_cuda_reduce_multi.cu [hunk 1]
```diff
@@ -164,6 +184,11 @@ ucc_status_t ucc_mc_cuda_reduce_multi(const void *src1, const void *src2,
             DT_REDUCE_INT(int64_t, op, src1, src2, dst, size, count, ld, stream,
                          bk, th);
             break;
+        case UCC_DT_FLOAT16:
+            ucc_assert(4 == sizeof(__half));
```
Comment (vspetrov) [inline src/components/mc/cuda/kernel/mc_cuda_reduce_multi.cu:188]: 4->2

**Review by vspetrov — CHANGES_REQUESTED** [review]
gtest?

File: src/components/mc/cuda/kernel/mc_cuda_reduce_multi.cu (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/components/mc/cuda/kernel/mc_cuda_reduce_multi.cu:188]: copy paste 🤦‍♂️

**Comment by Sergei-Lebedev** [comment]
> gtest?

it would not be easy to add, need conversion functions from fp16 to fp32 and back as well as some function that will check fp16 accuracy for fp32 numbers. There are intrinsics for some cpu arch but then we need config time detection for it. Overall it seems that test will be overcomplicated and we will need tests for test

**Comment by vspetrov** [comment]
> > gtest?
> 
> it would not be easy to add, need conversion functions from fp16 to fp32 and back as well as some function that will check fp16 accuracy for fp32 numbers. There are intrinsics for some cpu arch but then we need config time detection for it. Overall it seems that test will be overcomplicated and we will need tests for test

i meant for GPU, we do test mc reduce for GPU right? we can postpone it for later

**Review by vspetrov — APPROVED** [review]

**Review by manjugv — APPROVED** [review]

**Pull request merged**